# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile sum.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing sum.cpp


In [ ]:
!g++ sum.cpp -o sum
!./sum

5 9
14

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Overwriting output1.txt


In [ ]:
%%writefile input2.txt
-5 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
4

Writing output2.txt


In [ ]:
!./sum < input1.txt

12

In [ ]:
!./sum < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

4


In [ ]:
!grep -xvFf output2.txt output1.txt 

12


In [ ]:
!./sum < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./sum < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3


def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

1 5
6


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

/bin/bash: input2.txt: No such file or directory
grep: output2.txt: No such file or directory


# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

### Tipo de dato erroneo por lo que el dato no se puede guardar

In [19]:
%%writefile max_pairwise_product1.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product1.cpp


In [21]:
!g++ max_pairwise_product1.cpp -o max_pairwise_product1
!./max_pairwise_product1

6
23 1 89 56 34 5
4984


### Test 1: Small numbers

In [22]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [23]:
!./max_pairwise_product1 < input1.txt

90


### Test 2: lagest input numbers 200000 200000 in the input

In [24]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Writing input2.txt


In [25]:
!./max_pairwise_product1 < input2.txt

1345294336


### Código ineficiente pero con correcto tipo de dato

In [26]:
%%writefile max_pairwise_product2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product2.cpp


In [16]:
%%writefile output1.txt
40000000000

Writing output1.txt


In [27]:
!g++ max_pairwise_product2.cpp -o max_pairwise_product2
!./max_pairwise_product2 < input1.txt
!./max_pairwise_product2 < input2.txt

90
40000000000


In [28]:
!./max_pairwise_product2 < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt 

90


### Test 3 stress tests 20 times random n in [2, 20]

In [9]:
%%writefile Test3Write.cpp
#include <stdlib.h>
#include <time.h>
#include<iostream>
#include <vector>
#include <algorithm>
#include <fstream>
#include <string>
#include <cstdlib>
using namespace std;

int main(){
    
    for(int i=0;i<20;i++){
        string name = "Test3_input" + to_string(i+1) + ".txt";
        ofstream MyFile(name);
        int n = (rand() % 20) +2;
        MyFile << n <<"\n";
        srand(time(NULL));
        for(int c = 0; c < n; c++){
            MyFile << (rand() % 20000) + 1<< " ";;
        }
         MyFile.close();
    }
    return 0;
}

Overwriting Test3Write.cpp


In [10]:
!g++ Test3Write.cpp -o Test3Write
!./Test3Write

In [29]:
!./max_pairwise_product2 < Test3_input1.txt
!./max_pairwise_product2 < Test3_input2.txt
!./max_pairwise_product2 < Test3_input3.txt
!./max_pairwise_product2 < Test3_input4.txt
!./max_pairwise_product2 < Test3_input5.txt
!./max_pairwise_product2 < Test3_input6.txt
!./max_pairwise_product2 < Test3_input7.txt
!./max_pairwise_product2 < Test3_input8.txt
!./max_pairwise_product2 < Test3_input9.txt
!./max_pairwise_product2 < Test3_input10.txt
!./max_pairwise_product2 < Test3_input11.txt
!./max_pairwise_product2 < Test3_input12.txt
!./max_pairwise_product2 < Test3_input13.txt
!./max_pairwise_product2 < Test3_input14.txt
!./max_pairwise_product2 < Test3_input15.txt
!./max_pairwise_product2 < Test3_input16.txt
!./max_pairwise_product2 < Test3_input17.txt
!./max_pairwise_product2 < Test3_input18.txt
!./max_pairwise_product2 < Test3_input19.txt
!./max_pairwise_product2 < Test3_input20.txt

281608073
340592139
340592139
340592139
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073


### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [39]:
%%writefile Test4Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input4.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Writing Test4Write.cpp


In [40]:
!g++ Test4Write.cpp -o Test4Write
!./Test4Write

In [ ]:
!g++ max_pairwise_product2.cpp -o max_pairwise_product2
!time ./max_pairwise_product2 < input4.txt

39999800000

real	2m7.016s
user	2m6.812s
sys	0m0.021s


In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;     //1
    std::vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        std::cout << numbers[i];                                             
    }

    std::cout << "\n";                       
    return 0;                                                              
}

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product2.cpp


Count executed intructions in code and verify with inputs 1, 2 and 4

In [ ]:
%%writefile max_pairwise_product2_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Writing max_pairwise_product2_execints.cpp


In [ ]:
!g++ max_pairwise_product2_execints.cpp -o max_pairwise_product2_execints
!./max_pairwise_product2_execints < input1.txt
!./max_pairwise_product2_execints < input2.txt
!./max_pairwise_product2_execints < input4.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
89
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input4.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

In [ ]:
# Estimated time by instruction

128/80003672077

1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077)

625028688.1015625

### Solución eficiente pero con perdida de memoria

Add in each line a coment with the number of intstrcution that are executed in that line

In [30]:
%%writefile max_pairwise_product3.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max1 = 1;   // 1
	  long long max2 = 1;   // 1
    int n = numbers.size();   // 1

    for (int first = 0; first < n; first++) {   // 1 + 2n
        if(numbers[first]>max1){   // n
            max2 = max1;   // [0, n]
            max1 = (long long)numbers[first];   // [0, n]  
        }else if(numbers[first]>max2){   //[0, n]
            max2 =  (long long)numbers[first];   //[0, n]
        }
    }
    return max1*max2;       // 4 + 3n + [0, 2n]
}

int main() {
    int n;   // 1
    std::cin >> n;   // 1
    std::vector<int> numbers(n);   // n
    for (int i = 0; i < n; ++i) {   // 1 + 2n
        std::cin >> numbers[i];   // n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";   // 1 + 1 + 4 + 3n + [0, 2n]
    return 0;  
}             // 9 + 7n + [0, 2n]


Writing max_pairwise_product3.cpp


Test with inputs 1 and 2

In [31]:
!g++ max_pairwise_product3.cpp -o max_pairwise_product3
!./max_pairwise_product3 < input1.txt
!./max_pairwise_product3 < input2.txt

90
40000000000


Test and time it with input 3

In [ ]:
!time ./max_pairwise_product3 < input4.txt

39999800000

real	0m0.072s
user	0m0.064s
sys	0m0.003s


20 stress tests

In [32]:
!./max_pairwise_product3 < Test3_input1.txt
!./max_pairwise_product3 < Test3_input2.txt
!./max_pairwise_product3 < Test3_input3.txt
!./max_pairwise_product3 < Test3_input4.txt
!./max_pairwise_product3 < Test3_input5.txt
!./max_pairwise_product3 < Test3_input6.txt
!./max_pairwise_product3 < Test3_input7.txt
!./max_pairwise_product3 < Test3_input8.txt
!./max_pairwise_product3 < Test3_input9.txt
!./max_pairwise_product3 < Test3_input10.txt
!./max_pairwise_product3 < Test3_input11.txt
!./max_pairwise_product3 < Test3_input12.txt
!./max_pairwise_product3 < Test3_input13.txt
!./max_pairwise_product3 < Test3_input14.txt
!./max_pairwise_product3 < Test3_input15.txt
!./max_pairwise_product3 < Test3_input16.txt
!./max_pairwise_product3 < Test3_input17.txt
!./max_pairwise_product3 < Test3_input18.txt
!./max_pairwise_product3 < Test3_input19.txt
!./max_pairwise_product3 < Test3_input20.txt

281608073
340592139
340592139
340592139
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073


Count executed instructions in code and verify with inputs 1,2 and 3

In [33]:
%%writefile max_pairwise_product3_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {
    long long max1 = 1;   // 1
    execints++;

	  long long max2 = 1;   // 1
    execints++;   

    int n = numbers.size();   // 1
    execints++;

    execints++;
    for (int first = 0; first < n; first++) {   // 1 + 2n
        execints += 2;
        
        execints++;
        if(numbers[first]>max1){   // n
            max2 = max1;   // [0, n]
            execints++;

            max1 = (long long)numbers[first];   // [0, n]  
            execints++;

        }else if(numbers[first]>max2){   //[0, n]
            execints++;

            max2 =  (long long)numbers[first];   //[0, n]
            execints++;
        }else{
            execints++;
        }
    }
    
    return max1*max2;       // 4 + 3n + [n, 2n]
}

int main() {
    long long execints =0LL;

    int n;   // 1
    execints++;

    std::cin >> n;   // 1
    execints++;

    std::vector<int> numbers(n);   // n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {   // 1 + 2n
        execints += 2;

        std::cin >> numbers[i];   // n
        execints++;
    }

    // std::cout << MaxPairwiseProduct(numbers) << "\n";   // 1 + 1 + 4 + 3n + [n, 2n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout << "Numero total de instrucciones: " <<  execints << "\n"; 
    
    return 0;  
}             // 9 + 7n + [n, 2n]

Writing max_pairwise_product3_execints.cpp


In [ ]:
!g++ max_pairwise_product3_execints.cpp -o max_pairwise_product3_execints
!./max_pairwise_product3_execints < input1.txt
!./max_pairwise_product3_execints < input2.txt
!./max_pairwise_product3_execints < input4.txt

Numero total de instrucciones: 98
Numero total de instrucciones: 53
Numero total de instrucciones: 1800008


Calculate estimated time by instruction and number of instructions by second

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 18 = 98 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(5) + 9 = 53 $$

For the input4.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```
$$T = 9 + 7(200000) + 399999 = 1800008 $$

In [ ]:
# Estimated time by instruction

print(128/98)
print(128/53)
print(128/1800008)

1.3061224489795917
2.4150943396226414
7.111079506313305e-05


In [ ]:
# Estimated number of instructions by second

print(1/(128/98))
print(1/(128/53))
print(1/(128/1800008))

0.7656250000000001
0.4140625
14062.5625


### Solución más eficiente

Add in each line a coment with the number of intstrcution that are executed in that line

In [34]:
%%writefile max_pairwise_product4.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n;  // 1
    std::cin >> n;  // 1
    int number;  // 1
    long long max1 = 0;  // 1
	  long long max2 = 0;  // 1

    for (int i = 0; i < n; ++i) {  // 1 + 2n
        std::cin >> number;  // n
        if(number>max1){    // n
            max2 = max1;  // [0,n]
            max1 = (long long)number;  // [0,n]
        }else if(number>max2){  // [0,n]
            max2 =  (long long)number;  // [0,n]
        }
    }

    std::cout << max1*max2 << "\n"; // 1
    return 0;
}         // 7 + 4n + [n, 2n]


Writing max_pairwise_product4.cpp


Test with inputs 1 and 2

In [36]:
!g++ max_pairwise_product4.cpp -o max_pairwise_product4
!./max_pairwise_product4 < input1.txt
!./max_pairwise_product4 < input2.txt

90
40000000000


Test and time it with input 4

In [41]:
!./max_pairwise_product4 < input4.txt

39999800000


20 stress tests

In [37]:
!./max_pairwise_product4 < Test3_input1.txt
!./max_pairwise_product4 < Test3_input2.txt
!./max_pairwise_product4 < Test3_input3.txt
!./max_pairwise_product4 < Test3_input4.txt
!./max_pairwise_product4 < Test3_input5.txt
!./max_pairwise_product4 < Test3_input6.txt
!./max_pairwise_product4 < Test3_input7.txt
!./max_pairwise_product4 < Test3_input8.txt
!./max_pairwise_product4 < Test3_input9.txt
!./max_pairwise_product4 < Test3_input10.txt
!./max_pairwise_product4 < Test3_input11.txt
!./max_pairwise_product4 < Test3_input12.txt
!./max_pairwise_product4 < Test3_input13.txt
!./max_pairwise_product4 < Test3_input14.txt
!./max_pairwise_product4 < Test3_input15.txt
!./max_pairwise_product4 < Test3_input16.txt
!./max_pairwise_product4 < Test3_input17.txt
!./max_pairwise_product4 < Test3_input18.txt
!./max_pairwise_product4 < Test3_input19.txt
!./max_pairwise_product4 < Test3_input20.txt

281608073
340592139
340592139
340592139
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product4_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    long long execints =0LL;

    int n;  // 1
    execints++;

    std::cin >> n;  // 1
    execints++;

    int number;  // 1
    execints++;

    long long max1 = 1;  // 1
    execints++;

	  long long max2 = 1;  // 1
    execints++;

    execints++;
    for (int i = 0; i < n; ++i) {  // 1 + 2n
        execints+=2;

        std::cin >> number;  // n
        execints++;

        execints++;
        if(number>max1){    // n
            max2 = max1;  // [0,n]
            execints++;

            max1 = (long long)number;  // [0,n]
            execints++;

        }else if(number>max2){  // [0,n]
            execints++;

            max2 =  (long long)number;  // [0,n]
            execints++;

        }else{
            execints++;
        }
    }

    // std::cout << max1*max2 << "\n";

    execints++;
    std::cout << "Numero total de instrucciones: " <<  execints << "\n"; 
    return 0;
}     // 7 + 4n + [n, 2n]


Writing max_pairwise_product4_execints.cpp


In [ ]:
!g++ max_pairwise_product4_execints.cpp -o max_pairwise_product4_execints
!./max_pairwise_product4_execints < input1.txt
!./max_pairwise_product4_execints < input2.txt
!./max_pairwise_product4_execints < input4.txt

Numero total de instrucciones: 66
Numero total de instrucciones: 36
Numero total de instrucciones: 1200006


Calculate estimated time by instruction and number of instructions by second


In [ ]:
# Estimated time by instruction

print(128/66)
print(128/36)
print(128/1200006)

1.9393939393939394
3.5555555555555554
0.00010666613333599998


In [ ]:
# Estimated number of instructions by second

print(1/(128/66))
print(1/(128/36))
print(1/(128/1200006))

0.515625
0.28125
9375.046875


## Python

###Solucion ineficiente

En esta solucion no hay un tipo de dato que no corresponda, ya que python no maneja data types.

In [42]:
%%writefile max_pairwise_product1.py
# python3


def max_pairwise_product(numbers):
    n = len(numbers)  # 1
    max_product = 0 # 1
    for first in range(n): # n
        for second in range(first + 1, n): # n(n-1)/2
            max_product = max(max_product,
                numbers[first] * numbers[second])   # n(n-1)/2

    return max_product  # 2 + n + n(n-1)


if __name__ == '__main__':
    input_n = int(input())  # 1
    input_numbers = [int(x) for x in input().split()]   # n
    print(max_pairwise_product(input_numbers)) # 3 + 2n + n(n-1)


Writing max_pairwise_product1.py


### Test 1: Small numbers

In [ ]:
%%time
!python max_pairwise_product1.py < input1.txt

90
CPU times: user 5.43 ms, sys: 7.92 ms, total: 13.3 ms
Wall time: 128 ms


### Test 2: lagest input numbers 200000 200000 in the input

In [ ]:
%%time
!python max_pairwise_product1.py < input2.txt

40000000000
CPU times: user 5.92 ms, sys: 4.97 ms, total: 10.9 ms
Wall time: 118 ms


### Test 3 stress tests 20 times random n in [2, 20]

In [43]:
%%time
!python max_pairwise_product1.py < Test3_input1.txt
!python max_pairwise_product1.py < Test3_input2.txt
!python max_pairwise_product1.py < Test3_input3.txt
!python max_pairwise_product1.py < Test3_input4.txt
!python max_pairwise_product1.py < Test3_input5.txt
!python max_pairwise_product1.py < Test3_input6.txt
!python max_pairwise_product1.py < Test3_input7.txt
!python max_pairwise_product1.py < Test3_input8.txt
!python max_pairwise_product1.py < Test3_input9.txt
!python max_pairwise_product1.py < Test3_input10.txt
!python max_pairwise_product1.py < Test3_input11.txt
!python max_pairwise_product1.py < Test3_input12.txt
!python max_pairwise_product1.py < Test3_input13.txt
!python max_pairwise_product1.py < Test3_input14.txt
!python max_pairwise_product1.py < Test3_input15.txt
!python max_pairwise_product1.py < Test3_input16.txt
!python max_pairwise_product1.py < Test3_input17.txt
!python max_pairwise_product1.py < Test3_input18.txt
!python max_pairwise_product1.py < Test3_input19.txt
!python max_pairwise_product1.py < Test3_input20.txt

281608073
340592139
340592139
340592139
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
CPU times: user 149 ms, sys: 130 ms, total: 280 ms
Wall time: 2.41 s


### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [ ]:
%%time
!python max_pairwise_product1.py < input4.txt

Traceback (most recent call last):
  File "max_pairwise_product1.py", line 18, in <module>
    print(max_pairwise_product(input_numbers)) # 3 + 2n + n(n-1) + n * n(n-1)
  File "max_pairwise_product1.py", line 10, in max_pairwise_product
    numbers[first] * numbers[second])   # n * n(n-1)
KeyboardInterrupt
CPU times: user 23.1 s, sys: 3.15 s, total: 26.3 s
Wall time: 55min 32s


Esta prueba, al usar el algoritmo ineficiente, tiene un tiempo de ejcución muy largo, causando que no se muestre el resultado, por lo que no se va a analizar más adelante en esta sección.

####Count executed intructions in code and verify with inputs 1 and 2

In [ ]:
%%writefile max_pairwise_product1_execints.py

def max_pairwise_product(numbers, execints):
    n = len(numbers)  # 1
    execints += 1
    max_product = 0 # 1
    execints += 1

    m = 0

    for first in range(n): # n
        execints += 1
        for second in range(first + 1, n): # n(n-1)/2
            execints += 1
            m += 1
            max_product = max(max_product, numbers[first] * numbers[second])   # n(n-1)/2
            execints += 1
            m += 1
    #return max_product  # 2 + n + n(n-1) + n(n-1)/2
    print("Numero de instrucciones en la funcion: " + str(m))
    return execints


if __name__ == '__main__':
    execints = 0
    input_n = int(input())  # 1
    execints += 1
    input_numbers = [int(x) for x in input().split()]  # n
    execints = execints + input_n
    print("Numero de instrucciones totales: " + str(max_pairwise_product(input_numbers, execints))) # 3 + 2n + n(n-1)

Overwriting max_pairwise_product1_execints.py


In [ ]:
!python max_pairwise_product1_execints.py < input1.txt
!python max_pairwise_product1_execints.py < input2.txt

Numero de instrucciones en la funcion: 90
Numero de instrucciones totales: 113
Numero de instrucciones en la funcion: 30
Numero de instrucciones totales: 45


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 3 + 2n +n(n-1)$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 3 + 2(20) + (10)(9) = 113 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + (6)(5)  = 45 $$


El tiempo estimado se hace con el test 1 debido a que el input 4 no se puede compilar en este algoritmo.

In [ ]:
# Estimated time by instruction

0.128/113

0.0011327433628318584

In [ ]:
# Estimated number of instructions by second

1/(0.128/113)

882.8125

### Efficient solution 

In [44]:
%%writefile max_pairwise_product2.py
# python3

def max_pairwise_product(numbers):
  max1 = 0 # 1
  max2 = 0 # 1
  n = len(numbers) # 1

  for i in range(n): # n
    if(numbers[i] > max1): # n
      max2 = max1 # [0,n]
      max1 = numbers[i] # [0,n]
    elif(numbers[i] > max2): # [0,n]
      max2 = numbers[i] # [0,n]
    
  return max1 * max2 # 3 + 2n + [n,2n]

if __name__ == '__main__':
  input_n = int(input()) # 1
  input_numbers = [int(x) for x in input().split()] # n
  print(max_pairwise_product(input_numbers)) # 4 + 3n + [n,2n]

Writing max_pairwise_product2.py


### Test 1: Small numbers

In [ ]:
%%time
!python max_pairwise_product2.py < input1.txt

90
CPU times: user 7.78 ms, sys: 3.35 ms, total: 11.1 ms
Wall time: 121 ms


### Test 2: lagest input numbers 200000 200000 in the input

In [ ]:
%%time
!python max_pairwise_product2.py < input2.txt

40000000000
CPU times: user 6.91 ms, sys: 4.08 ms, total: 11 ms
Wall time: 118 ms


### Test 3 stress tests 20 times random n in [2, 20]

In [45]:
%%time
!python max_pairwise_product2.py < Test3_input1.txt
!python max_pairwise_product2.py < Test3_input2.txt
!python max_pairwise_product2.py < Test3_input3.txt
!python max_pairwise_product2.py < Test3_input4.txt
!python max_pairwise_product2.py < Test3_input5.txt
!python max_pairwise_product2.py < Test3_input6.txt
!python max_pairwise_product2.py < Test3_input7.txt
!python max_pairwise_product2.py < Test3_input8.txt
!python max_pairwise_product2.py < Test3_input9.txt
!python max_pairwise_product2.py < Test3_input10.txt
!python max_pairwise_product2.py < Test3_input11.txt
!python max_pairwise_product2.py < Test3_input12.txt
!python max_pairwise_product2.py < Test3_input13.txt
!python max_pairwise_product2.py < Test3_input14.txt
!python max_pairwise_product2.py < Test3_input15.txt
!python max_pairwise_product2.py < Test3_input16.txt
!python max_pairwise_product2.py < Test3_input17.txt
!python max_pairwise_product2.py < Test3_input18.txt
!python max_pairwise_product2.py < Test3_input19.txt
!python max_pairwise_product2.py < Test3_input20.txt

281608073
340592139
340592139
340592139
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
281608073
CPU times: user 123 ms, sys: 127 ms, total: 250 ms
Wall time: 2.35 s


### Test 4 time test measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [ ]:
%%time
!python max_pairwise_product2.py < input4.txt

39999800000
CPU times: user 6.24 ms, sys: 6.39 ms, total: 12.6 ms
Wall time: 215 ms


####Count executed intructions in code and verify with inputs 1, 2 and 4

In [ ]:
%%writefile max_pairwise_product2_execints.py
# python3

def max_pairwise_product(numbers, execints):
  max1 = 0 # 1
  execints += 1
  max2 = 0 # 1
  execints += 1
  n = len(numbers) # 1
  execints += 1
  m = 0
  for i in range(n): # n
    execints += 1
    

    execints += 1
    if (numbers[i] > max1): # n
      max2 = max1 # [0,n]
      execints += 1
      m += 1

      max1 = numbers[i] # [0,n]

      execints += 1
      m += 1
    elif (numbers[i] > max2): # [0,n]
      execints += 1
      m+= 1
      max2 = numbers[i] # [0,n]
      execints += 1
      m += 1

    else:
      execints += 1
      m += 1

  
  #return max1 * max2 # 3 + 2n + [n,2n]
  print("Numero de instrucciones en la funcion: " + str(m))
  return execints

if __name__ == '__main__':
  execints = 0
  input_n = int(input()) # 1
  execints += 1
  input_numbers = [int(x) for x in input().split()] # n
  execints += input_n
  print("Numero de instrucciones totales: " + str(max_pairwise_product(input_numbers, execints))) # 4 + 3n + [n,2n]

Overwriting max_pairwise_product2_execints.py


In [ ]:
!python max_pairwise_product2_execints.py < input1.txt
!python max_pairwise_product2_execints.py < input2.txt
!python max_pairwise_product2_execints.py < input4.txt

Numero de instrucciones en la funcion: 20
Numero de instrucciones totales: 54
Numero de instrucciones en la funcion: 11
Numero de instrucciones totales: 33
Numero de instrucciones en la funcion: 400000
Numero de instrucciones totales: 1000004


For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 4 + 3(10) + 20 = 54 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 4 + 3(6) + 11 = 53 $$

For the input4.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```
$$T = 4 + 3(200000) + 400000 = 1000004 $$

In [ ]:
# Estimated time by instruction

0.215/1000004

2.1499914000344e-07

In [ ]:
# Estimated number of instructions by second

1/(0.215/1000004)

4651181.395348837

# Screenhsoots of the Coursera Submissions in C++ and Python

![AplusB](https://raw.githubusercontent.com/jusantosm/AlgorithmsUN2021I/main/Lab%206/AplusB.JPG)

![MaxPairwiseProduct](https://raw.githubusercontent.com/jusantosm/AlgorithmsUN2021I/main/Lab%206/Max_Pairwise_Product.JPG)
